File: src/Path Searching Algorithm/Astar.py
import heapq
import math

# 휴리스틱 함수 : 유클리드 거리 계산 (노드 간의 직선 거리)
def heuristic(node1, node2):
    x1, y1 = node1
    x2, y2 = node2
    return math.sqrt((x1 - x2) **2 + (y1 - y2) **2)

def a_star(graph, start, goal):
    # 각 노드까지의 실제 거리 저장할 딕셔너리 (초기값 : 무한대)
    g_costs = {node: float('inf') for node in graph}
    g_costs[start] = 0

    # 우선순위 큐 초기화
    priority_queue = [(0, start)] # (f(n), node)
    came_from = {} # 경로 추적을 위한 딕셔너리

    while priority_queue:
        # 현재 노드를 큐에서 꺼냄
        current_f_cost, current_node = heapq.heappop(priority_queue)

        # 목표 노드에 도달한 경우 경로를 반환
        if current_node == goal:
            path = []
            total_cost = g_costs[goal]
            while current_node in came_from:
                path.append(current_node)
                current_node = came_from[current_node]
            path.append(start)
            return path[::-1], total_cost # 경로를 시작부터 끝까지로 반환
        
        # 인접 노드를 탐색
        for neighbor, cost in graph[current_node]:
            tentative_g_cost = g_costs[current_node] + cost
            if tentative_g_cost < g_costs[neighbor]:
                # 새로운 최단 거리 발견
                g_costs[neighbor] = tentative_g_cost
                f_cost = tentative_g_cost + heuristic(neighbor, goal)
                heapq.heappush(priority_queue, (f_cost, neighbor))
                came_from[neighbor] = current_node

    return None, None # 경로가 없는 경우

# 그래프 정의 (인접 리스트)
# 노드 위치가 좌표 (x, y)로 정의됨
graph = {
    (0, 0): [((2, 0), 3), ((0, 2), 2)],
    (2, 0): [((0, 0), 3), ((2, 2), 5), ((4, 0), 2)],
    (0, 2): [((0, 0), 2), ((2, 2), 4)],
    (2, 2): [((2, 0), 5), ((0, 2), 4), ((4, 2), 3)],
    (4, 0): [((2, 0), 2), ((4, 2), 1)],
    (4, 2): [((2, 2), 3), ((4, 0), 1), ((4, 4), 2)],
    (4, 4): [((4, 2), 2)]
}

start = (0, 0)
goal = (4, 4)
path, total_cost = a_star(graph, start, goal)

print("최단 경로 : ", path)
print("가중치 합 : ", total_cost)

File: src/Path Searching Algorithm/Astar.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <cmath>
#include <utility>
#include <limits>

using namespace std;

// 해시 함수 정의
struct hash_pair {
    template <class T1, class T2>
    size_t operator()(const pair<T1, T2>& p) const {
        auto hash1 = hash<T1>{}(p.first);
        auto hash2 = hash<T2>{}(p.second);
        return hash1 ^ (hash2 << 1);
    }
};

// 두 좌표 사이의 유클리드 거리 (휴리스틱)
double heuristic(pair<int, int> node1, pair<int, int> node2) {
    int x1 = node1.first, y1 = node1.second;
    int x2 = node2.first, y2 = node2.second;
    return sqrt((x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 -  y2));
}

// A* 알고리즘
pair<vector<pair<int, int>>, double> a_star(
    const unordered_map<pair<int, int>, vector<pair<pair<int, int>, int>>, hash_pair>& graph,
    pair<int, int> start, pair<int, int> goal) {
        unordered_map<pair<int, int>, double, hash_pair> g_costs;
        unordered_map<pair<int, int>, pair<int, int>, hash_pair> came_from;

        for (const auto& node : graph) {
            g_costs[node.first] = numeric_limits<double>::infinity();
        }
        g_costs[start] = 0;

    // 우선순위 큐 선언
    priority_queue<pair<double, pair<int, int>>, vector<pair<double, pair<int, int>>>, greater<pair<double, pair<int, int>>>> pq;
    pq.push({0, start});

    while (!pq.empty()) {
        auto [current_f_cost, current_node] = pq.top();
        pq.pop();

        if (current_node == goal) {
            vector<pair<int, int>> path;
            double total_cost = g_costs[goal]; // 최단 경로의 총 가중치
            while (came_from.find(current_node) != came_from.end()) {
                path.push_back(current_node);
                current_node = came_from[current_node];
            }
            path.push_back(start);
            reverse(path.begin(), path.end());
            return {path, total_cost};
        }

        for (const auto& [neighbor, cost] : graph.at(current_node)) {
            double tentative_g_cost = g_costs[current_node] + cost;
            if (tentative_g_cost < g_costs[neighbor]) {
                g_costs[neighbor] = tentative_g_cost;
                double f_cost = tentative_g_cost + heuristic(neighbor, goal);
                pq.push({f_cost, neighbor});
                came_from[neighbor] = current_node;
            }
        }
    }
    return {{}, numeric_limits<double>::infinity()}; // 경로가 없는 경우
}

int main() {
    // 그래프 정의
    unordered_map<pair<int, int>, vector<pair<pair<int, int>, int>>, hash_pair> graph;
    graph[{0, 0}] = {{{2, 0}, 3}, {{0, 2}, 2}};
    graph[{2, 0}] = {{{0, 0}, 3}, {{2, 2}, 5}, {{4, 0}, 2}};
    graph[{0, 2}] = {{{0, 0}, 2}, {{2, 2}, 4}};
    graph[{2, 2}] = {{{2, 0}, 5}, {{0, 2}, 4}, {{4, 2}, 3}};
    graph[{4, 0}] = {{{2, 0}, 2}, {{4, 2}, 1}};
    graph[{4, 2}] = {{{2, 2}, 3}, {{4, 0}, 1}, {{4, 4}, 2}};
    graph[{4, 4}] = {{{4, 2}, 2}};

    pair<int, int> start = {0, 0};
    pair<int, int> goal = {4, 4};
    auto [path, total_cost] = a_star(graph, start, goal);

    cout << "최단 경로 : \n";
    for (const auto& node : path) {
        cout << "(" << node.first << "," << node.second << ") ";
    }
    cout << "\n총 가중치 합 : " << total_cost << endl;

    return 0;
}